### DISCUSSION TITLE
геймплэй

### DISCUSSION THREAD

**Discussion by Pigzinzspace** [godot-engine-ru/shooter-3d#12] [Ideas]
**содержание документа** тут собраны общие темы, итоги обсуждения, предложения по любым частям игры которые не вошли в другие документы

**общее**
Игра однопользовательская. 
Предварительное рабочее название  Зомби Шутер: Мясорубка
Игра это динамичный 3D шутер, в котором будет много монстров и они будут разлетаться на части кровь и кишки. 
Монстры будут продолжать спавнится за спиной игрока, зачищенных участков на карте не будет.
Игра без драмы, ни у кого не гибнут дети, никто от ужаса не лезет в петлю, у нас бездумное мочилово где в монстрах нет ничего человеческого, бегаешь стреляешь, солнышко светит птички поют.
Действие первой части происходит внутри заводского  комплекса. 
Игра разделена на миссии, миссия это цель которой должен достичь игрок , иногда он ее знает иногда узнает по ходу самой миссии. 
В начале миссии показывается небольшая катсцена, после этого показывается текстовое описание миссии задание, в конце миссии показывают текстовый   брифинг, с очками, информацией по секретам, время прохождения, количество убитых зомби,  как в квэйк или дум. Катсцену, стартовый и финальный брифинг можно промотать нажатием любой клавиши.  

**игрок**
у игрока есть шкала HP, он умеет передвигаться, атаковать и взаимодействовать с окружением
игрок не умеет ускорятся
все оружие которое  использует игрок действует и на него.  
 
**оружие/предметы/инвентарь**
В игре может быть много оружия классического  и фантастического.
Игрок  может использовать 5 видов оружия, одновременно использовать можно по одному из каждой основной группы.
Расклад по оружию (это не позиции оружия которое есть в игре, конкретные позиции указаны в разделе "оружие гаджеты", тут приведены примеры оружия чтоб было понятно деление по слотам):
1 - мили оружие
2 - оружие короткой дистанции с пистолетными патронами или дробовики
3 - специфичесикое средства :  огнемет, гранатомет, ракетница
4 - снайперское, автоматические или полуавтоматические, нарезные и гладкоствольные, винтовки, карабины и автоматы
5 - какое то супер оружие, уникальное оружие 
6 - гранаты, ловушки,мины
У игрока нет инвентаря, все квестовые или другие предметы например ключи, огнеупорные костюмы, кусачки для вскрытия замка  применяются автоматически. 
Медикаменты применяются автоматически игрок встает на них или подходит к ним и они восстанавливают ему ХП мгновенно.
Патроны не бесконечные, но есть 4 вида патронов, патроны 2, патроны 3, патроны 4, патроны 5 они всегда подходят к тому оружию выбранной категории которое использует игрок.     
 
**монстры/мобы**
монстры используют атаки ближнего и дальнего боя, некоторые могут бегать по стенам и потолку
**спавн монстров** 
по комнатам, оп таймеру, и спавн ключевых монстров и босов
- по комнате
игрок подходит к комнате, в ней спавнятся мобы, это зависит только от комнаты, какая комната такие и мобы спавнятся, спавнятся 1 раз после загрузки комнаты, если комнату выгрузить потом опять загрузить заспавнятся снова
- круговой по времени 
спавнятся по таймеру со всех сторон от героя и бегут к нему сразу заагренные
- спавн босы и ключевых мобы
это отдельный тригер который срабатывает по сюжету
- максимальное расстояние от игрока
все зомби которые отстали от игрока запутались остались не заагренными в комнатах, если они дальше расстояния, немедленно удаляются
- верхний предел спавна
если вокруг игрока больше 40 зомби то перестает работать круговой спавн
если вокруг игрока 50 зомби то перестает работать  спавн в комнатах, спавнятся только  босы и ключевые монстры
- повторное использование 
модели зомби которые отстали или были убиты могут быть повтороно использованы. Если это нужно, они хранятся за спиной игрока или под полом там где игрок не может их увидеть, они деактивированы не издают звуки не используют скилы и не взаимодействуют пока они в таком состоянии.
- чекпоинты
 места где завершается миссия и начинается следующая, особые места в сюжете - приостанавливают таймер кругового спавна, чтоб миссия могла завершиться или ключевое событие могло произойти (например показать кат сцену), после того как события произошли чекпоинт теряет иммунитет к круговому спавну.      

**загрузка/сохраниение**
- Перед началом игры игрок выбирает свое имя
- Сохранение или загрузка в игре доступны только при завершении миссий, или на промежуточных чекпоинтах, если уровни большие по времени.  
- Условие для сохранения, - игрок должен выйти из боя. 
- Круговой спавн прекращается когда игрок приходит на активный чекпоинт и игрок может убить всех бегущих за ним мобов.
- Чекпоинты одноразовые, срабатывают автоматически и утрачивают иммунитет к круговому спавну. Нельзя вернуться чтоб мобы перестали спавнится.
- В сэйв сохряняется информация необходимая чтоб восстановить данные о состоянии игрока и игры на момент загрузки: хп, количество патронов каждого слота, количество гранат, типы оружия в каждом слоте, найденные игроком секреты и другая колючевая информация 

**оверлэй, анимация состояний**
изменение цвета экрана или специфический узор, капли крови, цепи, пламя по краям экрана:

- когда у игрока мало ХП,
- когда игрок получает урон от кислоты, токсинов, яда
- когда игрок получает урон от горения
- когда игрок по воздействием контроля - заморозки, оглушения, не стоит на ногах 

**окружение**
- есть коробки, бочки, ящики которые можно разрушать, некоторые из них взрываются 
- с некоторыми объектами можно взаимодействовать, например открыть дверь в шкафчик, перекусить душку замка, открыть холодильни, 
- взаимодействие иногда открывает меню выбора например выбор слота оружия.  

**локализация**
в игре будет локализация как минимум текстов ru/en которая выбирается в меню как максимум голосовой  озвучки

**Comment by slapin** [comment]
Нужно больше CC и тактического оружия. Например станящий газ или мины или разбрасываемые шипы, которые дают возможность выскочить из окружения / неприятных ситуаций. Должно быть в минимальных количествах доступно.

**Reply by Pigzinzspace** [reply]
фриз гранаты?
ослепляющие гранаты? (такие могут быть у охраны а фриз можно в лабе собрать)

**Reply by me2beats** [reply]
фриз гранаты норм, должно эффектно выглядеть. Замораживает — делает неподвижным на секунд 15-20. Стрельба во время заморозки размораживает.

Только он наверно и себя задеть может. Или сделать допущение что на самогО гг фриз не распространяется

**Reply by me2beats** [reply]
ослепляющие ещё проще сделать — зомби просто топчутся рандомно хаотично примерно на одном месте (+- несколько шагов)

**Reply by me2beats** [reply]
Мины можно
Шипы не так весело наверно и скорее всего сложнее реализовать

**Reply by Pigzinzspace** [reply]
> фриз гранаты норм, должно эффектно выглядеть. Замораживает — делает неподвижным на секунд 15-20. Стрельба во время заморозки размораживает.
> 
> Только он наверно и себя задеть может. Или сделать допущение что на самогО гг фриз не распространяется

да не 15-20 секунд это для шутера целая вечность, скорее 1 секунда а потом они еще секунд 5 двигаются как в слоумо приходя в норму

**Reply by Pigzinzspace** [reply]
> 

у тебя обычно нет возможности в шутере ставить ловушки, на тебя летит 10 мобов ты должен их из дробоша или кувалдой с трех шотов положить и добежать до двери пока не заспавнилось еще 10

**Reply by Pigzinzspace** [reply]
можно придумать какую то химическую ловушку с каким то запахом типо пакет крови об стенку разбиваешь, да точно и сделать их можно в офисе охраны в конце первой миссии, просто кровь из холодильника медпункта охраны взять

**Reply by slapin** [reply]
Шипы тоже достаточно просто, зомби в них влезает и топчется на одном месте.

**Reply by Pigzinzspace** [reply]
шипы это как то не реалистично

**Comment by me2beats** [comment]
Не уверен насчёт снайперского
У нас же небольшие расстояния на уровнях

**Reply by me2beats** [reply]
Хотя хотелось бы большие
Но годо вряд ли позволит

**Reply by slapin** [reply]
Ну пользу снайперки можно передать в уроне и точности. Но более медленном использовании.

**Reply by Pigzinzspace** [reply]
не обязательно чтоб можно было с любым оружием проходить игру

**Reply by slapin** [reply]
Интереснее когда оружие все-таки приходится иногда менять, но сильно мешать использовать любимое оружие тоже не стоит...

**Reply by Pigzinzspace** [reply]
> Интереснее когда оружие все-таки приходится иногда менять, но сильно мешать использовать любимое оружие тоже не стоит...

наверно стоит сделать еще одну тему организация тестов, только давай ты ее начнешь, ибо мне просто нечего там сказать
там надо описать как будут проводится  тесты (модели, анимация, эффекты я просто наобум пишу, вообще не в курсе тех процесса ) их этапы и что от  кого требуется для какого этапа  
типо тест №5
чтоб начать его нужно получить 
ответы на вопросы - 
код, модели, права, информацию -
делают его такие люди - 

как мы это будем делать

**Comment by me2beats** [comment]
что будем включать в себя сохранение?
Патроны/ здоровье /позиции заспавненных зомбей

**Reply by Pigzinzspace** [reply]
- В процессе прохождения миссии игрок сохраняться не может
Сохранение остаются только по миссиям, игрок не может сохранится в игре, а завершение миссии делает сохранение автоматически.

- Порядок сохранения/загрузки: 
Игрок регистрирует свой профиль - имя персонажа
Когда он проходит миссию игра запоминает его ХП, содержание оружейных слотов и гаджетов, количество патронов. И создает в его профиле загрузку с названием следующей миссии. 
Если игрок хочет выйти он теряет прогресс в не завершенной миссии.
Если надо прерваться игру можно поставить на паузу.

это нам даст что у нас будет столько начальных состояний в игре сколько миссий, а игроку добавит немного хардкора 
позицию зомби в таком случае сохранять не надо они генерятся по кругу и по комнатам.

**Reply by Pigzinzspace** [reply]
еще пару слов в защиту такого подхода
Профиль и авто сохранение только в начале миссии избавляет нас  от кучи проблем:
-запоминать расположение зомби, 
-запоминать выбранный слот оружия, кулдауны стрельбы, и используемых предметов. 
-летящие в игрока пули, 
-пассивность или агресивность мобов, 
-стадии атаки мобов, кулдауны их атак.
-текущий этап миссии  что уже сделано что еще нет, 
ведь нам при загрузке придется все это настроить.
Помните как сэйв/лоадом абузили проход в охраняемые, закрытые локации, или через агрессивных мобов например в готике 1, это только кажется простым куском, там очень много вещей придется учитывать.
Мы разрубаем этот гардиев узел запрещая загркзку в середине миссий, и делаем игру более напряженной, от начала миссии до конца миссии.

**Comment by slapin** [comment]
1. Слот запоминать нам и так не нужно
2. Пассивность/агрессивность - зачем это в сейве или это параметры миссии?
3. У нас слишком простая игра чтобы можно было что-то заабузить, но с
загрузками с начала миссии я согласен.
можно кстати сделать чекпоинты если миссии будут дольше 5 минут.

**Reply by Pigzinzspace** [reply]
> 2. Пассивность/агрессивность - зачем это в сейве или это параметры миссии?

на меня бежит 5 мобов которые меня точно убьют, я захожу за угол, сохраняюсь, загружаюсь, они меня не видят, они не агрессивны у меня есть время перезарядится подождать откат атаки и еще миллион способов

**Comment by slapin** [comment]
Вообще проблем сделать сохранение/загрузку всего в такой игре нет,
то есть если бояться багоюза то тут такой проблемы нету. Вопрос только в том
что если в середине миссии не будет сохранений, то она должна быть очень короткой.
Далеко не все будут счастливы перепроходить много раз по полчаса геймплея, задроты нынче не те.

**Reply by Pigzinzspace** [reply]
у нас миссии не длинные 10-15 минут 
через сейв проходится все что угодно на любой сложности 
это полностью лишает нас контроля над игрой и ее сложностью 
боятся что не всем зайдет не надо по тому что всегда заходит не всем 
мы выбрали формат герой бежит и косит толпы мобов и  мы должны игрока к такому геймплэю подталкивать
лучше ослабить мобов в целом и сделать отдельный хардкорный уровень сложности чем позволить игроку шланговать через сэйв

**Comment by slapin** [comment]
Со слотами все просто - каждый автосейв создает новый слот и пишет в него сейв при начале миссии. Количество слотов автосейва не огранично. При загрузке игры в меню Load появляется список слотов
1. 08:30am Jan 11 Название Миссии
2. 09:20am Jan 18 НазваниеМиссии2

И игрок может выбрать нужный слот нажав нужную кнопку. Так же игрок может удалить лишние слоты.

Профиль нам нужен если прохождения будут чем-то отличаться одно от другого, типа разными персонажами и тп.
Тогда мы в слоте указываем за какого персонажа играем, остальное не меняется.

**Reply by Pigzinzspace** [reply]
оно будет отличаться выбранным оружием, наличием патронов, оставшимся хп, найденными в игре секретами, про секреты забыли, хорошо что вспомнили

**Reply by Pigzinzspace** [reply]
> Тогда мы в слоте указываем за какого персонажа играем, остальное не меняется.

нам только на таком уровне профиль и нуже 
**имя,миссия**

**Comment by slapin** [comment]
А почему они вдруг станут не агрессивны? при нормальном AI это
невозможно. Но не суть, мы ж вроде договорились что промежуточное
состояние не сохраняем, только в конце миссии и на чекпоинтах.

**Reply by Pigzinzspace** [reply]
ну а откуда они узнают что были заагрены, это ведь какой то флаг в объекте который по умолчанию выключен, или координаты цели, они никого не видят, им надо сказать что они были заагрены, про это и речь

**Comment by slapin** [comment]
По расположению относительно цели они конечно узнают. Но конечно все что
является состоянием придется сохранить,
но мы ж решили что не будем. Мне в моем проекте вполне удается сохранить
все полностью благодаря архитектуре всего.
Но желание сделать по олдскулу могу понять.
